src/main/java/org/broadinstitute/hellbender/tools/exome/SampleDB.java

```diff
@@ -0,0 +1,429 @@
+ *     will follow the same lexicographical order.
+ * </p>
+ * <p>
+ *     "<i>Orphan</i>" read-groups, those that do not belong to any sample, are last, as if the "{@code null}"
+ *     sample's id is greater than any-other id in lexicographical order.
+ * </p>
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public final class SampleDB {
+
+    /**
+     * Total number of read-groups in this collection.
+     */
+    private final int readGroupCount;
+
+    /**
+     * Total number of samples in this collection.
+     */
+    private final int sampleCount;
+
+    /**
+     * Map from sample-id to its index <code>[0 .. {@link #sampleCount()}).</code>
+     */
+    private final Map<String,Integer> sampleIndexById;
+
+    /**
+     * Map from sample-id to its index <code>[0 .. {@link #sampleCount()}).</code>
+     */
+    private final Map<String,Integer> readGroupIndexById;
+    private final Map<String,Integer> sampleIndexByGroupId;
+    private final List<Sample> samples;
+    private final List<String> sampleIds;
+    private final List<String> readGroupIds;
+    private final int defaultSampleIndex;
+
+    public SampleDB(final SAMFileHeader header) {
+        this(header,null,true);
+    }
+
+    /**
+     * Create a new sample-db from the read file header.
+     *
+     * <p>
+     *     You can indicate a default sample id to be used for read-groups without a sample id or for reads with no
+     *     read-group.
+     * </p>
+     *
+     * <p>
+     *     If you pass a {@code null} default sample id, subsequent calls to {@link #sampleIndexByRead(SAMRecord)} will
+     *     return {@code -1} on reads without a sample.
+     * </p>
+     *
+     * <p>
+     *     Even when you indicate a default sample id you can tell the constructor to exclude it from the final
+     *     sample collection if there is no sample-id less read-group and you don't expect to see read-group less reads.
+     * </p>
+     *
+     * @param header the input read source header.
+     * @param defaultSampleId the default sample id for orphan reads or read-groups.
+     * @param ignoreDefaultSampleWithoutOrphanReadGroups whether we should exclude the default sample
+     *                                                   if there is no sample-id less read-group.
+     *
+     * @throws IllegalArgumentException if {@code header} is {@code null}.
+     *  is reserved for internal use.
+     */
+    public SampleDB(final SAMFileHeader header, final String defaultSampleId, final boolean ignoreDefaultSampleWithoutOrphanReadGroups) {
+        if (header == null) {
+            throw new IllegalArgumentException("the header provided cannot be null.");
+        }
+        final List<SAMReadGroupRecord> readGroupRecords = header.getReadGroups();
+        final int readGroupCount = readGroupRecords.size();
+
+        final Set<String> orphanReadGroups = new HashSet<>(readGroupCount);
```

Comment:
Supports for sample-id less read-groups. They can be assigned a default sample-id. Not currently using the functionality although is fully unit-tested. Do you want me to remove it?